pkg/client/tx/client_test.go

```diff
@@ -446,7 +451,153 @@ func TestTxClient_SignAndBroadcast_Timeout(t *testing.T) {
+		expectedTx cometbytes.HexBytes
+		// txResultsBzPublishChMu is a mutex that protects txResultsBzPublishCh from concurrent access
+		// as it is expected to be updated in a mock method but is also sent on in the test.
+		txResultsBzPublishChMu = new(sync.Mutex)
+		// txResultsBzPublishCh is the channel that the mock events query client
+		// will use to publish the transactions event bytes. It is not used in
+		// this test but is required to use the NewOneTimeTxEventsQueryClient
+		// helper.
+		txResultsBzPublishCh chan<- either.Bytes
+		// blocksPublishCh is the channel that the mock block client will use
+		// to publish the latest block. It is not used in this test but is
+		// required to use the NewOneTimeCommittedBlocksSequenceBlockClient
+		// helper.
+		blocksPublishCh chan client.Block
+		ctx             = context.Background()
+	)
+
+	keyring, signingKey := testkeyring.NewTestKeyringWithKey(t, testSigningKeyName)
+
+	// Construct a new mock events query client. Since we expect the
+	// NewTxClient call to fail, we don't need to set any expectations
+	// on this mock.
+	eventsQueryClient := testeventsquery.NewOneTimeTxEventsQueryClient(
+		ctx, t, signingKey, txResultsBzPublishChMu, &txResultsBzPublishCh,
+	)
+
+	// Instruct the tx client to return an error when submitting a transaction.
+	callStatus := &callStatus{
+		errorToReturn: fmt.Errorf("test error"),
+	}
+	txCtxMock := newTxContext(t, ctx,
+		testSigningKeyName, keyring, &expectedTx, callStatus,
+	)
+
+	// Construct a new mock block client because it is a required dependency.
+	// Since we're not exercising transactions timeouts in this test, we don't need to
+	// set any particular expectations on it, nor do we care about the contents
+	// of the latest block.
+	blockClientMock := testblock.NewOneTimeCommittedBlocksSequenceBlockClient(
+		t, blocksPublishCh,
+	)
+
+	// Construct a new depinject config with the mocks we created above.
+	txClientDeps := depinject.Supply(
+		eventsQueryClient,
+		txCtxMock,
+		blockClientMock,
+	)
+
+	// Construct the transaction client.
+	txClient, err := tx.NewTxClient(
+		ctx, txClientDeps, tx.WithSigningKeyName(testSigningKeyName),
+	)
+	require.NoError(t, err)
+
+	// Construct an arbitrary message to sign, encode, and broadcast.
+	signingAddr, err := signingKey.GetAddress()
+	require.NoError(t, err)
+
+	appStake := types.NewCoin(volatile.DenomuPOKT, math.NewInt(1000000))
+	appStakeMsg := &apptypes.MsgStakeApplication{
+		// Providing address to avoid panic from #GetSigners().
+		Address:  signingAddr.String(),
+		Stake:    &appStake,
+		Services: client.NewTestApplicationServiceConfig(testServiceIdPrefix, 1),
+	}
+
+	// Sign and broadcast the message.
+	go txClient.SignAndBroadcast(ctx, 0, appStakeMsg)
+
+	// Wait for 5 seconds to allow the retry strategy to perform 4 failing retries.
+	time.Sleep(5 * time.Second)
+
+	// All attempts should have failed.
+	require.Equal(t, 4, callStatus.errorCount)
```

Comment:
I'm a little worried about potential CI flakinss here.

Have you thought and/or looked into it?

Reply:
Yes I did. The subject under test is by nature very time related but it showed no signs of flakiness.

The exponential retry params and sleep time is in the order of `seconds` so no signs of flakiness should be observed. I'll keep an eye on it though.